### ISSUE TITLE
Unexpected type mismatch

### ISSUE THREAD

**Issue by IanManske** [nushell/nushell#9825]
### Describe the bug

In the reproduction below, the `$split` variable is being stored with a type of `String` instead of `List(List(String))` in the engine state. As a result, this causes type checking to report an error when there shouldn't be one. Part of the issue seems to stem from the usage of `$in`. If `$in` is replaced with a regular closure input variable, then this issue is avoided.

### How to reproduce

```nushell
let text = "a b c\nd e f"
let split = ($text | lines | each { $in | split row ' ' })
$split | skip 1
```
gives:
```
Error: nu::parser::input_type_mismatch

  × Command does not support string input.
   ╭─[entry #1:2:1]
 2 │ let split = ($text | lines | each { $in | split row ' ' })
 3 │ $split | skip 1
   ·          ──┬─
   ·            ╰── command doesn't support string input
   ╰────
```
`$split`'s actual type and contents:
```nushell
> $split | describe 
list<list<string>>

> $split
╭───┬───────────╮
│ 0 │ ╭───┬───╮ │
│   │ │ 0 │ a │ │
│   │ │ 1 │ b │ │
│   │ │ 2 │ c │ │
│   │ ╰───┴───╯ │
│ 1 │ ╭───┬───╮ │
│   │ │ 0 │ d │ │
│   │ │ 1 │ e │ │
│   │ │ 2 │ f │ │
│   │ ╰───┴───╯ │
╰───┴───────────╯
```
The issue is avoided if there is no `$in`:
```nushell
> let split = ($text | lines | each {|line| $line | split row ' ' })

> $split | skip 1
╭───┬───────────╮
│ 0 │ ╭───┬───╮ │
│   │ │ 0 │ d │ │
│   │ │ 1 │ e │ │
│   │ │ 2 │ f │ │
│   │ ╰───┴───╯ │
╰───┴───────────╯
```
The following works as well, since there is no intermediate variable. This is despite `$in` being used.
```nushell
$text | lines | each { $in | split row ' ' } | skip 1
```

### Expected behavior

The type of `$split` should be tracked properly so that there is no type mismatch.

### Configuration

| key                | value                                                            |
| ------------------ | ---------------------------------------------------------------- |
| version            | 0.83.0                                                           |
| branch             | makepkg                                                          |
| commit_hash        | a33b5fe6ce97b5e9fe8a774c13e783ed65c1b591                         |
| build_os           | linux-x86_64                                                     |
| build_target       | x86_64-unknown-linux-gnu                                         |
| rust_version       | rustc 1.71.0 (8ede3aae2 2023-07-12) (Arch Linux rust 1:1.71.0-1) |
| cargo_version      | cargo 1.71.0                                                     |
| build_time         | 2023-07-26 01:10:19 +00:00                                       |
| build_rust_channel | release                                                          |
| allocator          | standard                                                         |
| features           | dataframe, default, extra, sqlite, trash, which, zip             |
| installed_plugins  |                                                                  |

### Additional context

I was able to reproduce this issue on latest main too. This issue is not present on v0.82.0.

**Comment by sholderbach** [comment]
Thanks for the report! This goes on our list of things to fix with #9812

**Comment by IanManske** [comment]
Looks like this will be fixed by #9827
